processing/src/main/java/org/apache/druid/java/util/common/guava/ParallelMergeCombiningSequence.java

```diff
@@ -1188,11 +1231,50 @@ boolean isCancelled()
 
     RuntimeException getRuntimeException()
     {
-      Throwable ex = throwable.get();
-      if (ex instanceof RuntimeException) {
-        return (RuntimeException) ex;
+      return wrapRuntimeException(throwable.get());
+    }
+
+    private static RuntimeException wrapRuntimeException(Throwable t)
+    {
+      if (t instanceof RuntimeException) {
+        return (RuntimeException) t;
       }
-      return new RE(ex);
+      return new RuntimeException(t);
+    }
+  }
+
+  public static class CancellationFuture extends AbstractFuture<Boolean>
```

Comment:
I considered this, but `AbstractFuture` is quite a lot more complicated than the simple `AtomicReference` of the `CancellationGizmo` with a variety of synchronization things going on internally to keep everything thread-safe. This, along with the listeners being triggered when stuff happens, i felt a bit uneasy about any of the threads in the pool being able to trigger this directly, so left them split up for now.

The future is basically just so it can participate in the external query cancellation stuff, so it seemed safer for it to only exist on the outside sequence, which runs on the http request thread.